### PR TITLE
added default `CacheAdapter` lifecycle hook

### PR DESCRIPTION
**Pull request by zilto**
This follows a similar logic to `hamilton.plugins.h_diskcache`, but uses the standard `shelve` library. It has the same API as the diskcache, but provides less utilities to manage cache size.

## Changes
- added a `CacheAdapter` to `hamilton.lifecycle.default`
- added relevant tests, derived from `tests/plugins/h_diskcache`

## How I tested this
- ran tests successfully locally

## Notes
- one edge case regarding cache history is mentioned in TODOs, but it shouldn't affect usability.

## Checklist

- [x] PR has an informative and human-readable title (this will be pulled into the release notes)
- [x] Changes are limited to a single goal (no scope creep)
- [x] Code passed the pre-commit check & code is left cleaner/nicer than when first encountered.
- [x] Any _change_ in functionality is tested
- [x] New functions are documented (with a description, list of inputs, and expected output)
- [x] Placeholder code is flagged / future TODOs are captured in comments
- [x] Project documentation has been updated if adding/changing functionality.

### CODE REVIEW

**Comment by skrawcz** [comment]
@zilto this should have a companion example, and also docs entry.

File: hamilton/lifecycle/default.py [hunk 1]
```diff
@@ -313,6 +316,109 @@ def run_after_node_execution(
             pdb.set_trace()
 
 
+class CacheAdapter(NodeExecutionHook, NodeExecutionMethod, GraphExecutionHook):
+    """Class to cache node results on disk with a key baed on the node code implementation and inputs.
```
Comment (elijahbenizzy) [inline hamilton/lifecycle/default.py:320]:
```suggestion
    """Class to cache node results on disk with a key based on the node code implementation and inputs.
```

File: hamilton/lifecycle/default.py [hunk 2]
```diff
@@ -313,6 +316,109 @@ def run_after_node_execution(
...
+    ):
+        """Create cache key based on node callable hash (equiv. to HamiltonNode.version) and
+        the node inputs (`node_kwargs`).If key in cache (cache hit), load result; else (cache miss),
+        compute the node and append node name to `used_nodes_hash`.
+
+        Note:
+            - the callable hash is stored  in `used_nodes_hash` because it's required to create the
+            key in `run_after_node_execution` and the callable won't be accessible to recompute it
+        """
+        if node_name not in self.cache_vars:
```
Comment (elijahbenizzy) [inline hamilton/lifecycle/default.py:368]: For afterwards -- we may want to make this a set? Odds are high this is not a slow operation, but it could get `O(n**2)` which we don't want.

**Review by elijahbenizzy — APPROVED** [review]
Looks good! A few comments to fix after, merging.
